### PR TITLE
Unit test client, worker and coordinator code

### PR DESCRIPTION
**Pull request by petroav**
* The Script() and Execute() classes were mocked with actual
  modules because
    1) they are dependencies of the code under test
    2) they do not exist at 'compile time' and are only
       available at runtime on the cluster
    3) those classes are deeply embedded in the main Ambari
       repo and have not been extracted into their own wheel/egg
       to be declared as dependencies of this project
    4) it is impossible/prohibitively expensive to mock the modules
       in the test code
* Some of these unit tests are a little meaningless and will
  break with a refactor. I included them because no other
  tests (other than manual) are possible and they provide a basic
  sanity check.

Task: SWARM-1125
Testing: make clean dist test

### CODE REVIEW

**Comment by petroav** [comment]
@anusudarsan and @kokosing, when you get a chance I would appreciate comments

File: MANIFEST.in [hunk 1]
```diff
@@ -10,4 +10,4 @@ recursive-include configuration *
 
 recursive-exclude * __pycache__
 recursive-exclude * *.py[co]
-recursice-exclude tests *
\ No newline at end of file
+recursice-exclude tests *
```
Comment (anusudarsan) [inline MANIFEST.in]: s/recursice/recursive

File: tests/test_cli.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+class TestCli(unittest.TestCase):
+
+    @patch('package.scripts.presto_cli.Execute')
+    def test_install_shells_out_to_execute(self, execute_mock):
+        presto_cli = Cli()
+
+        presto_cli.install(MagicMock)
+
+        assert execute_mock.called
```
Comment (anusudarsan) [inline tests/test_cli.py]: do we want to assert if execute is called with certain args?

Comment (petroav) [inline tests/test_cli.py]: We could, but I tried to minimize the fragility of the unit test by just asserting that we're shelling out. Otherwise the unit tests will fail if we change the command we execute.

File: package/scripts/presto_client.py [hunk 3]
```diff
@@ -80,7 +80,8 @@ def ensure_nodes_are_up(client):
         raise RuntimeError('Presto server failed to start within \
 {0} seconds.'.format(RETRY_TIMEOUT))
 
-
+# This class was copied more or less verbatim from
```
Comment (anusudarsan) [inline package/scripts/presto_client.py]: where are you using this module? sorry I dont exactly remember when you added  this

File: tests/test_presto_client.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        self.presto_client_mock.get_rows.return_value = TestPrestoClientSmoketest.nation
+
+        smoketest_presto(self.presto_client_mock)
+
+        assert self.presto_client_mock.execute_query.called
+
+    def test_failure_when_fewer_than_25_rows_returned_from_nation(self):
+        self.presto_client_mock.get_rows.return_value = [1]
+
+        TestCase.assertRaises(self, RuntimeError, smoketest_presto, self.presto_client_mock)
```
Comment (anusudarsan) [inline tests/test_presto_client.py]: how is this assertRaises different from the assertRaisesRegexp function at line 202?

File: package/scripts/presto_client.py (hunk 3 above)
Comment (petroav) [inline package/scripts/presto_client.py]: The line right below shows you where I got this code from. I use it in the smoketest function, which is at the top of this file. That function accepts a PrestoClient() instance as an argument and goes on to validate the installation. In turn the smoketest function is called from the Coordinator class (in the presto_coordinator.py file) in the start method.

Comment (anusudarsan) [inline package/scripts/presto_client.py]: got it :) thanks

File: tests/test_cli.py (hunk 2 above)
Comment (anusudarsan) [inline tests/test_cli.py]: I think that will adding the arg check will help us from any minor regressions..eg: a very simple typo..anyway, your call.

File: package/scripts/presto_client.py (hunk 3 above)
Comment (anusudarsan) [inline package/scripts/presto_client.py]: just curious, why are you not checking in the ensure_* methods above in this file for the exact number of nodes or exact type of connectors which are active?

**Comment by anusudarsan** [comment]
lgtm mod comments.

File: MANIFEST.in (hunk 1 above)
Comment (petroav) [inline MANIFEST.in]: Done.

File: package/scripts/presto_client.py (hunk 3 above)
Comment (petroav) [inline package/scripts/presto_client.py]: Yes that is the next step in the integration. I need to figure out how to extract the topology information the user sets during setup and pass it down to these functions.

File: tests/test_presto_client.py (hunk 4 above)
Comment (petroav) [inline tests/test_presto_client.py]: assertRaises will only check that an exception of the given type was raised whereas assertRaisesRegexp will actually check the error message as well

**Comment by petroav** [comment]
Thanks for the review, I'll wait for @kokosing to take a look tomorrow as well.

File: tests/test_presto_client.py (hunk 4 above)
Comment (anusudarsan) [inline tests/test_presto_client.py]: ok..just checking if you can replace this one with raisesRegexp..nvm

File: package/scripts/presto_client.py [hunk 5]
```diff
@@ -80,7 +80,8 @@ def ensure_nodes_are_up(client):
...
+# https://github.com/prestodb/presto-admin/blob/master/prestoadmin/prestoclient.py
```
Comment (kokosing) [inline package/scripts/presto_client.py]: It would be nice to avoid copying code. Is it possible to upload prestoadmin to some python repositories and use presto-admin as dependencies?

File: tests/resources/valid_rest_response_level1.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+{"id":"2015_harih","infoUri":"http://localhost:8080/v1/query/2015_harih","nextUri":"http://localhost:8080/v1/statement/2015_harih/2"}
```
Comment (kokosing) [inline tests/resources/valid_rest_response_level1.txt]: maybe add indentation?

File: tests/test_cli.py (hunk 2 above)
Comment (kokosing) [inline tests/test_cli.py]: no new line ;)

File: tests/test_coordinator.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+        with patch('__builtin__.open', open_mock):
+            presto_coordinator.configure(self.mock_env)
+
+        assert create_tpch_connector_mock.called
+
+    @patch('package.scripts.params.config_properties', new=dummy_config_properties)
+    @patch('package.scripts.presto_coordinator.smoketest_presto')
+    @patch('package.scripts.presto_coordinator.Coordinator.configure')
+    @patch('package.scripts.presto_coordinator.Execute')
```
Comment (kokosing) [inline tests/test_coordinator.py]: can you group this patches? So you just use one here and above?

File: tests/test_coordinator.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+        presto_coordinator.status(self.mock_env)
+        assert execute_mock.called
+
+    @patch('package.scripts.params.config_properties', new=dummy_config_properties)
+    @patch('package.scripts.presto_coordinator.smoketest_presto')
+    @patch('package.scripts.presto_coordinator.Coordinator.configure')
+    @patch('package.scripts.presto_coordinator.Execute')
+    def test_install_start_configure_presto(
+            self, unused_execute_mock, configure_mock, unused_smoketest_presto):
```
Comment (kokosing) [inline tests/test_coordinator.py]: if mocks are unused, why do you provide them?

File: tests/test_coordinator.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+        presto_coordinator.stop(self.mock_env)
+        assert execute_mock.called
+        execute_mock.reset_mock()
+
+        presto_coordinator.start(self.mock_env)
+        assert execute_mock.called
+        execute_mock.reset_mock()
+
+        presto_coordinator.status(self.mock_env)
+        assert execute_mock.called
```
Comment (kokosing) [inline tests/test_coordinator.py]: don't you need to check other mocks that they were unused?

File: tests/test_coordinator.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+    @patch('package.scripts.params.config_properties', new=dummy_config_properties)
+    @patch('package.scripts.presto_coordinator.smoketest_presto')
+    @patch('package.scripts.presto_coordinator.Coordinator.configure')
+    @patch('package.scripts.presto_coordinator.Execute')
+    def test_lifecycle_methods_shell_out_to_execute(
+            self, execute_mock, unused_configure_mock, unused_smoketest_presto):
+        presto_coordinator = Coordinator()
+
+        presto_coordinator.install(self.mock_env)
+        assert execute_mock.called
```
Comment (kokosing) [inline tests/test_coordinator.py]: I think you need to check some arguments here, because install can do the same as stop and test will pass.

I would suggest to check only some arguments (maybe just one) to not to make test too fragile. So I would check execute_mock was called with something related to install for presto_coordinator.install, related to stop for presto_coordinator.stop etc

File: tests/test_presto_client.py [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+
+    @patch('package.scripts.presto_client.ensure_nodes_are_up', side_effect=RuntimeError())
+    def test_failure_when_nodes_are_not_up(self, ensure_nodes_are_up_mock):
+        TestCase.assertRaises(self, RuntimeError, smoketest_presto, self.presto_client_mock)
+
+    @patch('package.scripts.presto_client.ensure_catalogs_are_available', side_effect=RuntimeError())
+    def test_failure_when_catalogs_are_not_available(self, ensure_catalogs_are_available_mock):
+        TestCase.assertRaises(self, RuntimeError, smoketest_presto, self.presto_client_mock)
```
Comment (kokosing) [inline tests/test_presto_client.py]: import assertRaises

File: tests/test_worker.py [hunk 12]
```diff
@@ -12,27 +12,112 @@
...
+
+    @patch('package.scripts.presto_worker.create_tpch_connector')
+    @patch('package.scripts.params.config_properties', new=dummy_config_properties)
+    def test_config_properties_coordinator_always_false(self, create_tpch_connector_mock):
+        config = []
+        open_mock = mock_file_descriptor_write_method(config)
+        presto_worker = Worker()
+
+        with patch('__builtin__.open', open_mock):
+            presto_worker.configure(self.mock_env)
```
Comment (kokosing) [inline tests/test_worker.py]: above 5 lines of code is repeated in several tests, maybe extract this as method?

**Comment by kokosing** [comment]
looks good. A lot of mocking... hard to follow, but I understand it was necessary.

File: tests/test_cli.py (hunk 2 above)
Comment (petroav) [inline tests/test_cli.py]: Good call, done.

File: package/scripts/presto_client.py (hunk 5 above)
Comment (petroav) [inline package/scripts/presto_client.py]: Yea it's possible to extract the Python presto client. However, it would be a non trivial amount of work. As I was writing this code I realized that anyway we should release the presto Python client as a separate project for data scientists and such. So at some point we'll do it but right now I'm focused on completing this integration.

File: tests/resources/valid_rest_response_level1.txt (hunk 6 above)
Comment (petroav) [inline tests/resources/valid_rest_response_level1.txt]: Done.

File: tests/test_coordinator.py (hunk 8 above)
Comment (petroav) [inline tests/test_coordinator.py]: They are needed because otherwise the real method gets called. In this case I want my mock smoketest_presto method to be called instead of the real one.

File: tests/test_coordinator.py (hunk 9 above)
Comment (petroav) [inline tests/test_coordinator.py]: Well they are used :) I need them there so that the real method doesn't get called. When I say "unused", I mean not tested.

File: tests/test_coordinator.py (hunk 10 above)
Comment (petroav) [inline tests/test_coordinator.py]: I thought about it a little bit and you're right, I've been persuaded. It's a little silly that install can do the same as stop and the test will still pass. I'm adding changes that will allow verify one keyword in the argument.

File: tests/test_presto_client.py (hunk 11 above)
Comment (petroav) [inline tests/test_presto_client.py]: I don't think that's possible. I tried a couple of import combinations but nothing worked.

File: tests/test_worker.py (hunk 12 above)
Comment (petroav) [inline tests/test_worker.py]: Done. I did the same in test_coordinator.py.
